pkg/services/datasources/service/datasource.go

```diff
@@ -252,6 +261,162 @@ func (s *Service) AddDataSource(ctx context.Context, cmd *datasources.AddDataSou
 	})
 }
 
+// This will valid validate the instance settings and mutate the cmd with the processed values
+func (s *Service) prepareAdd(ctx context.Context, cmd *datasources.AddDataSourceCommand) error {
+	operation := backend.InstanceSettingsOperation_CREATE
+
+	if cmd.Type == "" {
+		return nil // This happens in tests
+	}
+
+	// Make sure it is a known plugin type
+	p, found := s.pluginStore.Plugin(ctx, cmd.Type)
+	if !found {
+		return fmt.Errorf("plugin '%s' not found", cmd.Type)
+	}
+
+	// When the APIVersion is set, the client must also implement ProcessInstanceSettings
+	if p.APIVersion == "" {
+		if cmd.APIVersion != "" {
+			return fmt.Errorf("invalid request apiVersion (datasource does not have one configured)")
+		}
+		return nil
+	}
+
+	jd, err := cmd.JsonData.ToDB()
+	if err != nil {
+		return fmt.Errorf("invalid jsonData (%v)", operation)
+	}
+
+	rsp, err := s.pluginClient.ProcessInstanceSettings(ctx, &backend.ProcessInstanceSettingsRequest{
+		PluginContext: backend.PluginContext{
+			OrgID:    cmd.OrgID,
+			PluginID: cmd.Type,
+			DataSourceInstanceSettings: &backend.DataSourceInstanceSettings{
+				UID:                     cmd.UID,
+				Name:                    cmd.Name,
+				URL:                     cmd.URL,
+				Database:                cmd.Database,
+				Updated:                 time.Now(),
+				JSONData:                jd,
+				DecryptedSecureJSONData: cmd.SecureJsonData,
+				Type:                    cmd.Type,
+				User:                    cmd.User,
+				BasicAuthEnabled:        cmd.BasicAuth,
+				BasicAuthUser:           cmd.BasicAuthUser,
+				APIVersion:              cmd.APIVersion,
+			},
+		},
+		Operation:   operation,
+		CheckHealth: false, // we have never checked in the past, should we start?
+	})
+	if err != nil {
+		if errors.Is(err, plugins.ErrMethodNotImplemented) {
+			return fmt.Errorf("plugin (%s) with apiVersion=%s must implement ProcessInstanceSettings", p.ID, p.APIVersion)
+		}
+	}
+	if rsp == nil {
+		return fmt.Errorf("expected response (%v)", operation)
+	}
+	settings := rsp.DataSourceInstanceSettings
+	if !rsp.Allowed || settings == nil {
+		return fmt.Errorf("not allowed (%+v)", rsp.Result)
+	}
+
+	// Use the mutated values
+	cmd.APIVersion = settings.APIVersion
+	cmd.BasicAuth = settings.BasicAuthEnabled
+	cmd.BasicAuthUser = settings.BasicAuthUser
+	cmd.URL = settings.URL
+	cmd.Database = settings.Database
+	cmd.SecureJsonData = settings.DecryptedSecureJSONData
+	cmd.JsonData = nil
+	if settings.JSONData != nil {
+		cmd.JsonData = simplejson.New()
+		err := cmd.JsonData.FromDB(settings.JSONData)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// identical to prepareAdd -- but the types do not overlap :(
+func (s *Service) prepareUpdate(ctx context.Context, cmd *datasources.UpdateDataSourceCommand) error {
```

Comment:
With https://github.com/grafana/grafana/pull/88018 -- we won't have to duplicate this complicated logic!